### PR TITLE
[CLD-7750] Introducing Plugin signing with GitHub Actions

### PR DESCRIPTION
**Pull request by phoinixgrr**
#### Summary

Introducing: 

- Migrating plugin signing to GitHub Actions https://github.com/mattermost/actions-workflows/pull/45/commits/ca39b78d58c05e4a29b844de9c4bd563fed43cfa
-  Fix Steps naming related to CD process https://github.com/mattermost/actions-workflows/commit/c91320723826fa3a10f8c1099dce0173e9e3d262

Review by commit is recommended. 

#### Ticket Link

Ticket: https://mattermost.atlassian.net/browse/CLD-7750

### CODE REVIEW

**Comment by phoinixgrr** [comment]
What are the benefits of that approach? The deliverable here is the signing of plugins, which is being successfully achieved. This github-action releases the actual plugins, with an additional signing step added. 

I see no reason to increase complexity by utilizing another repository (delivery platform) and the Argo-events platform, which would introduce dependencies in another repository, require CI on a different repository, and utilize AWS infrastructure to achieve this.

**Review by mvitale1989 — CHANGES_REQUESTED** [review]
Thanks Akis :rocket:  I see a possible cleanup for the gpg invocations, but other than that LGTM.

Regarding the where should the plugin release logic be located, I understand Saturn's point about keeping it consistent.
But I would be in favour of leaving things as they are, since all the mechanisms are already in place across all plugin repos. A migration would bring consistency, but I think it'd be better to dedicate our time on other higher priority tasks instead.

**Comment by mvitale1989** [comment]
Also a general process question: do we need to upload the plugin and signature anywhere else (e.g. S3) as part of this task, to fully supersede matterbuild? I'm wondering how for example does the plugin eventually make its way to the Marketplace.

File: .github/workflows/plugin-cd.yml
Comment (phoinixgrr) [reply .github/workflows/plugin-cd.yml]: Thanks for the comment @mvitale1989 🚀 

I am in agreement with the comment here. We can maybe push this for migration to the argo-event platform, as a an improvement, not relevant to this epic. 👍 

Concerning the    ` --no-default-keyring` option. 
This is actually the same functionality as already being used in the `plugin-signer` instance script, and is mostly transfered here. Not a GPG expert, but as I understand this has no actual effect in the process. 

Removed and rebased.

**Comment by phoinixgrr** [comment]
> Also a general process question: do we need to upload the plugin and signature anywhere else (e.g. S3) as part of this task, to fully supersede matterbuild? I'm wondering how for example does the plugin eventually make its way to the Marketplace.



> Also a general process question: do we need to upload the plugin and signature anywhere else (e.g. S3) as part of this task, to fully supersede matterbuild? I'm wondering how for example does the plugin eventually make its way to the Marketplace.

Plugin-signer instance, only uploads the signatures to Github Releases. This functionality is replicated here.

**Review by saturninoabril — APPROVED** [review]
Cool, then I’m good as is. Great work @phoinixgrr!

Thoughts: I don't think it's adding complexity because the resources are already in place on idle waiting to consume, nor asking to migrate for this is a change commented to realign with the current approach. I can see we have a really good foundation on structure and if this (gpg or release in general in the delivery-platform) is one of those and practical to do so, then I'd suggest being consistent. Having consistent/documented structure will help others (Devs) perform their job (self-serve) without depending on us much as we look forward to scale with them in this area.

**Comment by phoinixgrr** [comment]
> Cool, then I’m good as is. Great work @phoinixgrr!
> 
> Thoughts: I don't think it's adding complexity because the resources are already in place on idle waiting to consume, nor asking to migrate for this is a change commented to realign with the current approach. I can see we have a really good foundation on structure and if this (gpg or release in general in the delivery-platform) is one of those and practical to do so, then I'd suggest being consistent. Having consistent/documented structure will help others (Devs) perform their job (self-serve) without depending on us much as we look forward to scale with them in this area.

Thank you for commenting Saturn. 

The complexity is introduced in the following manner: 
- Introducing dependencies across various systems and platforms such as K8s, AWS, Argo-Events, AWS Load Balancers, Webhooks etc 
- The resulting code to consume the event in Argo-Events would be around 100 lines, whereas we achieve the same task here in just 5 lines.
- CI for release will break in a new repository delivery-platform - multiple places to host code
- The CI output will remain consistent and complete within the Plugin repository, instead of being transferred to another repository for the release process.
- More time and resources required for debugging and resolving issues that arise from the complex setup.
- Possibility for reduced performance and increased latency to deliver due to the added layers of dependencies and integrations.

As discussed, I am not dropping the proposal of moving this to Argo-Events for the sake of consistency. IMHO it is unnecessary, as we are only using the platform (argo-events) for the sake of using it and in the same time introducing complexity and dependencies, However, I understand the point about consistency.

Migrating the release process of plugins to the Argo-Events ecosystem for the sake of consistency should not be considered part of signing Mattermost plugins in GitHub

**Review by toninis — COMMENTED** [review]
Super 🚀 
Only 1 things for clarification:  
This pipeline alone does not deprecate matterbuild since it is also used by devs to  create the tag. We need to instruct developers to do it locally from now on or introduce Makefile commands for ease of use. 

Nice to see simplified pipelines 🚀

File: .github/workflows/plugin-cd.yml
Comment (toninis) [inline .github/workflows/plugin-cd.yml:131]: `private_key.asc` file is not under dist/ directory . This will probably fail . 
Need to remove the `working-directory` directive

Comment (phoinixgrr) [reply .github/workflows/plugin-cd.yml]: fixed 👍

**Comment by phoinixgrr** [comment]
> Super 🚀 Only 1 things for clarification: This pipeline alone does not deprecate matterbuild since it is also used by devs to create the tag. We need to instruct developers to do it locally from now on or introduce Makefile commands for ease of use.

Well known @toninis . This does not finish this migration. 
Plans are to include Makefile Targets and relevant documentation as already did in other software projects. Example https://github.com/mattermost/mattermost-push-proxy/pull/116 



> Nice to see simplified pipelines 🚀

<3 <3 <3 Glad that we are aligned

**Comment by phoinixgrr** [comment]
cc: @iyampaul  For security review.

**Comment by phoinixgrr** [comment]
There is a requirement, that Releases should be performed on private pipelines and infra. 

Do we have any input, on how's this release process not aligning to this policy? These are public pipelines that release mattermost plugins. 

It is out of the scope of the signing artifacts epic. , but maybe we should take this chance and migrate everything (release + signing of plugins) to argo-events, to align with this policy.
This can surely justify the added complexity in order to use the argo-events platform.

cc: @saturninoabril WDYT?

**Comment by saturninoabril** [comment]
I see, that's what I'm looking for, found the [referenced design docs](https://docs.google.com/document/d/1ibkYI7AYXyePOh2uI548zF5e3hgeR3PAQkuKd_B2wHE/edit#heading=h.crt3y328k3sr) and [hard requirements posted in channel](https://hub.mattermost.com/private-core/pl/g538sunbd7dqpxq7r358bngidw). I'll defer to Security but I think we need to realign. Yes, our supported plugins are in public repos/pipelines like with most of our projects but we do release in private.

**Comment by phoinixgrr** [comment]
Increasing the scope of the epic. Will transfer the release pipeline of plugins to argo-events, and private pipelines, for security reasons. Complexity introduced, justifies this reasoning. 

Closing this.

**Pull request closed without merging**
